File: euler191.py
"""
To compute the number of possible strings of length n, we define recursive function T(n) which
counts the number of prize strings of length n containing only the characters A and O. For n >= 3,
we have three options for the start of the string: 'AAO', 'AO', or 'O', each of which able to
subsequently branch into any of the three. For n <= 2, any string is possible so we have:

T(n) = { 2^n                        if n <= 2
       { T(n-1) + T(n-2) + T(n-3)   else

When there is an L at position i, the total number of strings is the product of the number of
possibilities before the L and the number after. We sum over all possible positions of L, and the
situation where there are no Ls to get the result.
"""

R = {} # dict for memoization
# T(n): counts number of prize strings containing O, A of length n
def T(n):
    if n not in R:
        R[n] = 2**n if n < 3 else T(n-1) + T(n-2) + T(n-3)
    return R[n]

N = 30 # default case
# N = 30000 # can handle up to around here
[T(n) for n in range(N)] # do bottom up DP to avoid exceeding recursion depth
print(T(N) + sum(T(i) * T(N-1-i) for i in range(N)))

File: euler191.cpp
// non dynamic solution in c++

#include <iostream>
#include <cmath>

using namespace std;

// counts prize strings of n length with previous characters c0, c1
int count_substrings_helper(char c0, char c1, int n) {
    int count = 0;
    if (n > 0) {
        // can always have next character be 'O'
        count += count_substrings_helper(c1, 'O', n - 1);

        // if last two characters aren't 'AA', then can have next char be 'A'
        if (!(c0 == 'A' && c1 == 'A'))
            count += count_substrings_helper(c1, 'A', n - 1);
    } else {
        count += 1;
    }
    return count;
}

// count number of prize strings with only 'A', 'O' characters and length n
int count_substrings(int n) {
    return count_substrings_helper('O', 'O', n);
}

// counts number of prize strings of length n
int count_prizestrings(int n) {
    int count = 0;

    for (int i = 0; i < n; i++) {
        int j = n - 1 - i;
        count += count_substrings(i) * count_substrings(j);
    }

    count += count_substrings(n);
    return count;
}

int main() {
    int n = 30;

    int count = count_prizestrings(n);
    cout << count << endl;
}
